src/gt4py/backend/dace_backend.py

```diff
@@ -366,13 +365,17 @@ def unique_index(self) -> int:
 
     def generate_entry_params(self, stencil_ir: gtir.Stencil, sdfg: dace.SDFG) -> List[str]:
         res: Dict[str, str] = {}
-        import dace.data
 
         for name in sdfg.signature_arglist(with_types=False, for_call=True):
             if name in sdfg.arrays:
                 data = sdfg.arrays[name]
                 assert isinstance(data, dace.data.Array)
-                res[name] = "py::buffer {name}, std::array<gt::int_t,{ndim}> {name}_origin".format(
+                res[
+                    name
+                ] = "py::{pybind_type} {name}, std::array<gt::int_t,{ndim}> {name}_origin".format(
+                    pybind_type="object"
+                    if self.backend.storage_info["device"] == "gpu"
+                    else "buffer",
```

Comment:
Just curious: why do these need to be different? I'm sure there's a good reason. I'm not too familiar with the details of pybind11.

Reply:
They should have not been `py::buffer` even now. It works now, since the GPU storages are both a `py::buffer` (due to the dual CPU buffer)  but also have the `__cuda_array_interface__` attribute that `as_cuda_sid` accesses. However, now we pass `CuPy` arrays from the python side, so the old interface breaks down.